### PR TITLE
Universal/stencil override

### PR DESCRIPTION
**Pull request by kaychang-unity**
---
### Purpose of this PR
This adds support for stencil override for deferred renderer.

---
### Testing status

**Manual Tests**: What did you do?
- [ X ] Opened test project + Run graphic tests locally with deferred renderer
- [ X ] Built a player

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fstencil_override

---
### Comments to reviewers

This code relies on C++ code in trunk that has not been backported to 2020.1 branch.
Should we #ifdef some of it (ScriptableRenderContext.DrawRenderers() )?

### CODE REVIEW

**Comment by kaychang-unity** [comment]
@alindmanUnity I made some changes in ShaderGraph so that it is possible to set custom shaderTags. Would it be better to implement a more general mechanism to specify a list of shader tags {name,value} pairs? We could then express RenderType values with that mechanism.

**Comment by kaychang-unity** [comment]
Sure, "UniversalMaterialType" is a custom tag for URP.

**Comment by alindmanUnity** [comment]
> @alindmanUnity I made some changes in ShaderGraph so that it is possible to set custom shaderTags. Would it be better to implement a more general mechanism to specify a list of shader tags {name,value} pairs? We could then express RenderType values with that mechanism.

I'm not opposed to it, I would sync with Antoine or Seb to see if that would be at all useful for HDRP as well. If not, this system is fine as is. Let me know if you want help to check with HD generation as well, and don't forget to run the HD tests to confirm no halo errors

**Comment by kaychang-unity** [comment]
@alindmanUnity how do test HD generation? Just run HDRP test suites as usual?

**Comment by kaychang-unity** [comment]
@oleks-k Is there anything you need?

**Comment by kaychang-unity** [comment]
@phi-lira If a shaderPass with GBuffer pass doesn't declare the material type tag, stencil state will not be set and it will not be shaded by the additional lights, as it will be considered unlit/baked lit type of materials.
For the documentation part, @oleks-k has been added to this PR.

**Comment by phi-lira** [comment]
Since Lit is our default material / shader, imo not defining a material type tag should behave as Lit as well.

**Comment by sebastienlagarde** [comment]
Hey, I have been thinking about this after Kay-chang talked about this PR.
My worries is that right now the system work because URP use a minimal set of bits in stencil. So you end with few combination for your tag. However, with time and more feature (in particular decal, SSR etc...), you will start to use more stencil bit and you will need more tag. For HDRP using the tag system required 32 variant of the stencil settings. Mean a loop of 32 for every objects to looks for stencil state. This is not optimal.

What about instead providing the stencil mask we want to apply to the pass?

Both URP and HDRP have a concept of users reserved stencil bit. What we should do instead of adding a custom tag option to DrawRenderer is to add an option specifically for this cases of reserved bit.

Lets have something like

DrawRenderer(....,  bool enableStencilUsersMak, StencilUserMask, StencilUserRef).

When enableStencilUsersMak is true, we combine any stencil state of Material with the one from the user.
i.E C++ Side:

StencilState

Mask = enableStencilUsersMak ? (currentMask & StencilUserMask) : currentMask ; // Enabled users bits
Ref = enableStencilUsersMak ? (currentRef | StencilUserRef) : currentRef ; 

This is cleaner from my point of view and more efficient.

Thought?

**Comment by sebastienlagarde** [comment]
Sorry to be the pain in the ass so late in the process, but lot of thing happening everywhere...

File: com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs [hunk 1]
```diff
@@ -24,12 +26,31 @@ public GBufferPass(RenderPassEvent evt, RenderQueueRange renderQueueRange, Layer
...
+                );
+            }
+
+            m_ShaderTagValues = new ShaderTagId[3];
+            m_ShaderTagValues[0] = new ShaderTagId("Lit");
+            m_ShaderTagValues[1] = new ShaderTagId("SimpleLit");
+            m_ShaderTagValues[2] = new ShaderTagId("Unlit");
+
+            m_RenderStateBlocks = new RenderStateBlock[3];
+            m_RenderStateBlocks[0] = OverwriteStencil(m_RenderStateBlock, 96, 32);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs:51]: I really suggest that you create a StencilUsage class for stencilm bit, it is quickly unreadable and you can make a lot of mistake: https://github.com/Unity-Technologies/Graphics/blob/0161c9151e11f854f472cbf7c6903bbb7c90083c/com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDStencilUsage.cs

Also users need an API to access stencil bits from C# side when they do custom pass in C#.

Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs:51]: Sure, I can replace these hardcoded constants.

**Comment by kaychang-unity** [comment]
@sebastienlagarde I am ok with this:
`DrawRenderers(...., bool enableStencilUsersMask, StencilUserMask, StencilUserRef). // of course we should use a struct or something cleaner)
`
Probably the structure would be GfxStencilState or DeviceStencilState. At the moment, the user can not only change the stencilRef but also actual stencil functions.

But I would like to keep setting stencil state centrally (via C# code) for the material types (or whatever we use stencil for in the future), as it avoid code duplication and is more maintainable long term (also this would prevent users from hard-coding our constants in their shader code).

Could this still be done in a different PR though, as adding a new implementation for DrawRenderers() will probably take another 1-2 weeks before it lands.

**Comment by sebastienlagarde** [comment]
> Probably the structure would be GfxStencilState or DeviceStencilState. At the moment, the user can not only change the stencilRef but also actual stencil functions.
> 

Hmm... How do you expect thing to work at all if users can change stencil function? It mean you have no guarantee that your Material bits will be written, or I missed something?

> But I would like to keep setting stencil state centrally (via C# code) for the material types (or whatever we use stencil for in the future), as it avoid code duplication and is more maintainable long term (also this would prevent users from hard-coding our constants in their shader code).
> 

Avoiding to have Material bits everywhere is a good thing, I am agree. But have in mind that it will come sooner than later that you will require a bit per Material (i.e not based on the Materia type but a bit for any kind of Material). Very random example, TexMesh pro ask for one bit to do some stuff in their shader, and a TextMeshPro can be Lit, unlit or simple Lit. 

Then you will need to combine the Material bit of TextMesh pro with the predefined bit of your Material and the one for users.
Of course, it works it is doable, but it still mean that you will need to declare the text Mesh pro bit in your shader.

The way we deal with it in HDRP is that we setup stencil in Lit.shader Like this:

            Stencil
            {
                WriteMask [_StencilWriteMaskGBuffer]
                Ref [_StencilRefGBuffer]
                Comp Always
                Pass Replace
            }

And _StencilWriteMaskGBuffer is setup by the inspector.

That's way we can drive the bit by our stencil usage class. If we do a change, then the Material upgrader do a call to SetupBaseUnlitKeywords() that we have define for all our Material and it will update the stencil bit to the correct changed value.

If a users create a Material in C#, he then need to call SetupBaseUnlitKeywords() explicitly (not done yet in HDRP, it is internal).

I will not say it is ideal, and it is correct that you can keep your Material centric C# bit, combine with the users one with the new API I propose (by considering Material bits are like users bits) and combine with the Material one.

I let you think what is best for you, as you know what you want to achieve, I am just highlighting that you shouldn't see shortem (i.e try to consider what will be your future feature, as once done and deliver. If you don't have all the upgrade Material mechanism setup, changing something will be impossible).

> Could this new implementation be done in a different PR though, as adding a new implementation for DrawRenderers() will probably take another 1-2 weeks before it lands.

This is why I haven't block the PR. I let you judge how you want to deal with those information and what is the best for URP from your point of view. If you feel confident about this PR, do it. But URP this year will come with tons of feature. I really advice to consider them and not shot yourself in the foot (or rather, shoot the users because you will need to do adjustment after this).

Nevertheless, I am still fully agree on your willing to have all stencil control from C# and nothing in .shader. This is a nice goal. It was not practical for HDRP due to various optimization we wanted to perform, but for URP it may be achievable.

**Comment by kaychang-unity** [comment]
> Hmm... How do you expect thing to work at all if users can change stencil function?
A subset of stencil states is valid. In essence, as long as the user allow writing stencil ref, then it can work. I added extra code to check if the user pass "valid" stencil states and some hover strings in the UI to explain requirements for the deferred renderer. I assume if they are going to use custom stencil tests outside the default ("Aways"), they have some understanding how it works.

Thanks for all the information regarding how HDRP does it. You only reserve 2bits stencil for user, correct?

At the moment, I would like to land this PR. I know we are already running out of stencil bits though. At least the new solution is more maintainable.

**Comment by phi-lira** [comment]
@sebastienlagarde thanks for your comments. I like the approach of combining the user and built-in stencil mask/ref in instead of looping material tags. This would also play well with our RendererFeature construct. 

`DrawRenderer(...., bool enableStencilUsersMak, StencilUserMask, StencilUserRef)` 
I don't see why we need the `enableStencilUserMask` here if we already have a specific overload for it. If we only have these two stencil parameters I think creating a struct is overshoot (duplication with other stencil related structs, more API, more docs).

**Comment by phi-lira** [comment]
@kaychang-unity my concern is if we ship the material tag for users and then want to change to another approach then we will break them. I guess if we are confident we will fix this before leaving preview it's fine but we should have a plan. How about we sync about it tomorrow?

File: com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs (hunk 1 above)
Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs]: Refactored into our own StencilUsage enum.

**Comment by kaychang-unity** [comment]
@phi-lira 
> Since Lit is our default material / shader, imo not defining a material type tag should behave as Lit as well.
I committed a small change to make Lit the default if UniveralMaterialType is not defined or the value does not match anything known.

File: com.unity.render-pipelines.universal/Editor/DeferredRendererDataEditor.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@ private static class Styles
             public static readonly GUIContent preferDepthPrepassLabel = EditorGUIUtility.TrTextContent("Prefer Depth Prepass", "Some platform configurations may still enable depth-prepass even when this checkbox is disabled.");
             public static readonly GUIContent accurateGbufferNormalsLabel = EditorGUIUtility.TrTextContent("Accurate G-buffer normals", "normals in G-buffer use octaedron encoding/decoding (expensive)");
             public static readonly GUIContent tiledDeferredShadingLabel = EditorGUIUtility.TrTextContent("Tiled Deferred Shading (Experimental)", "Allows Tiled Deferred Shading on appropriate lights");
+            public static readonly GUIContent invalidStencilOverride = EditorGUIUtility.TrTextContent("Deferred renderer reserves the 4 highest bits of stencil to store material types. Stencil Overrides settings are constrained to ensure correct lighting results.\n");
```
Comment (oleks-k) [inline com.unity.render-pipelines.universal/Editor/DeferredRendererDataEditor.cs:22]: @kaychang-unity , I need help in understanding this text. I'll ping you on Slack.

**Comment by phi-lira** [comment]
Merge master again to fix CI issues.

**Pull request merged**